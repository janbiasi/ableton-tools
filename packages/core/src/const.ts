import { homedir } from 'os';
import { resolve } from 'path';

export const ENCODING = 'utf8';

export const HOME_DIR = homedir();

export const ABLETON_TOOLS_ROOT = resolve(HOME_DIR, '.abletontools');

export const VERSION = require('../package.json').version;
