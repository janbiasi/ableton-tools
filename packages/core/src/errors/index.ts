import { BackupRestorationError } from './BackupRestorationError';
import { InvariantViolationError } from './InvariantViolationError';
import { ProjectNotFoundError } from './ProjectNotFoundError';
import { ProjectMetaNotFoundError } from './ProjectMetaNotFoundError';
import { DatabaseError } from './DatabaseError';

export {
	BackupRestorationError,
	InvariantViolationError,
	ProjectNotFoundError,
	ProjectMetaNotFoundError,
	DatabaseError,
};
