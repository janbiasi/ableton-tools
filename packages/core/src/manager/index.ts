export { Manager } from './Manager';
export { TemplateManager } from './TemplateManager';
export { ProjectManager } from './ProjectManager';
