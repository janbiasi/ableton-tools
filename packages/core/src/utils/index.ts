import { findLiveSetsInFolder } from './als';
import { createArchive, unpackArchive } from './compression';
import { fileExists, folderExists, getLastPathFragment } from './fs';
import { invariant } from './invariant';
import { safeDirName, formatBytes } from './string';
import { uuid } from './uuid';

export { findLiveSetsInFolder, createArchive, unpackArchive, fileExists, folderExists, getLastPathFragment, invariant, safeDirName, formatBytes, uuid };
