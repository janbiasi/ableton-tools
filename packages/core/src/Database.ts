import * as fs from 'fs-extra';
import { resolve } from 'path';
import { DatabaseError } from './errors';
import { uuid, fileExists } from './utils';
import { ENCODING, ABLETON_TOOLS_ROOT } from './const';

export interface IDatabaseConfig {
	name: string;
	root?: string;
	fileName?: string;
}

export interface IDatabaseEntry<T> {
	id: string;
	timestamp: number;
	data: T;
}

export interface IDatabaseContent<T extends { [key: string]: any }> {
	name: string;
	entries: IDatabaseEntry<T>[];
}

export class Database<T extends { [key: string]: any }> {
	private contents: IDatabaseContent<T> = {
		name: 'unknown',
		entries: [],
	};

	constructor(private config: IDatabaseConfig) {}

	public async init() {
		this.contents = {
			name: this.config.name,
			entries: [],
		};

		// ensure base directory exists
		await fs.mkdirp(this.root);

		// get initial database contents and safe DB initially
		await this.syncFromDatabase();
		await this.syncToDatabase();
	}

	public get entries() {
		return this.contents.entries as Readonly<IDatabaseEntry<T>[]>;
	}

	public async reset() {
		this.contents = {
			name: this.contents.name,
			entries: [],
		};

		await this.syncToDatabase();
	}

	/**
	 * Update or create an entry
	 */
	public async save(entry: T) {
		const index = this.contents.entries.push({
			id: uuid(),
			timestamp: Date.now(),
			data: entry,
		});

		await this.syncToDatabase();

		return this.contents.entries[index - 1];
	}

	public async deleteByID(id: string) {
		const entry = this.getById(id);

		if (entry) {
			let deletedItem: IDatabaseEntry<T> = {} as IDatabaseEntry<T>;

			this.contents.entries = this.entries.reduce((prev, entry) => {
				const itemOrNothing = entry.id === id ? [] : [entry];
				entry.id === id ? (deletedItem = entry) : void 0;

				return [...prev, ...itemOrNothing];
			}, [] as IDatabaseEntry<T>[]);

			await this.syncToDatabase();

			return deletedItem;
		}

		return false;
	}

	/**
	 * Generic getter for entries using a custom lookup function
	 */
	public get(lookup: (entry: IDatabaseEntry<T>) => boolean): IDatabaseEntry<T>[] {
		return this.contents.entries.filter(lookup);
	}

	/**
	 * Generic getter for one entry using a custom lookup function
	 */
	public getOne(lookup: (entry: IDatabaseEntry<T>) => boolean): IDatabaseEntry<T> | undefined {
		const [entry] = this.get(lookup);
		return entry;
	}

	/**
	 * Get an entry by ID
	 */
	public getById(id: string): IDatabaseEntry<T> | undefined {
		return this.contents.entries.find((entry): boolean => entry.id === id);
	}

	public async syncToDatabase() {
		try {
			await fs.writeFile(this.databaseFile, this.serialize());
		} catch (err) {
			throw new DatabaseError(this.config.name, `Sync to database has failed: ${err.message}`);
		}
	}

	public async syncFromDatabase() {
		const doesDatabaseExist = await this.checkDatabaseExists();

		if (!doesDatabaseExist) {
			// no database found, return current state
			return;
		}

		try {
			// database found, try reading and persisting it
			const dehydrated = await fs.readFile(this.databaseFile, ENCODING);
			this.contents = this.deserialize(dehydrated);
		} catch (err) {
			throw new DatabaseError(this.config.name, `Could not read database: ${err.message}`);
		}
	}

	public toString() {
		return `Database(${this.config.name}) { ${this.contents.entries.length} }`;
	}

	private get root() {
		return this.config.root || ABLETON_TOOLS_ROOT;
	}

	private get databaseFile() {
		return resolve(this.root, `${this.config.fileName || this.config.name}.json`);
	}

	private async checkDatabaseExists() {
		return await fileExists(this.databaseFile);
	}

	private serialize() {
		return JSON.stringify(this.contents, null, 2);
	}

	private deserialize(contents: string): IDatabaseContent<any> {
		return JSON.parse(contents);
	}
}
