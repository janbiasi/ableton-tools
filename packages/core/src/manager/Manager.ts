import fs from 'fs-extra';
import { resolve } from 'path';
import { ABLETON_TOOLS_ROOT, ENCODING } from '../const';
import { safeDirName } from '../utils/string';
import { invariant } from '../utils/invariant';
import { EventEmitter } from 'events';
import { fileExists, folderExists } from '../utils/fs';

/**
 * Abstract manager class to provide basic managing functionality
 * for example saving global configurations, template lists etc.
 */
export abstract class Manager {
	public prepared = false;
	public events = new EventEmitter();

	protected root: string;

	/**
	 * Creates a new manager with a name (base folder) and a possible
	 * override option for the global root, eg. if you want to save
	 * all your data in a different root directory than the $HOME folder.
	 */
	constructor(public name: string, globalRoot = ABLETON_TOOLS_ROOT) {
		this.root = resolve(globalRoot, safeDirName(name));
	}

	/**
	 * Generates a human readable name for the current manager
	 */
	public toString() {
		return `Manager { ${this.name} }`;
	}

	/**
	 * Every manager needs to implement an init method,
	 * you're required to first call "this.prepare()"!
	 */
	abstract init(): Promise<void>;

	/**
	 * Function for initialization, ensures the base folder and connections
	 */
	protected async prepare() {
		await fs.mkdirp(this.root);
		this.prepared = true;
	}

	/**
	 * Ensure that a certain subfolder exists for the current manager
	 */
	protected async ensureFolder(path: string) {
		invariant(this.prepare, `Ensure to call "this.prepare" in ${this.name}.init() implementation`);
		await fs.mkdirp(path);
	}

	/**
	 * Write a file to the current manager space
	 */
	protected async writeFile(name: string, content: object) {
		invariant(this.prepare, `Ensure to call "this.prepare" in ${this.name}.init() implementation`);
		await fs.writeFile(resolve(this.root, `${name}.json`), JSON.stringify(content));
	}

	/**
	 * Read a file from the current manager space
	 */
	protected async readFile<T extends object = {}>(name: string) {
		invariant(this.prepare, `Ensure to call "this.prepare" in ${this.name}.init() implementation`);
		const contents = await fs.readFile(resolve(this.root, `${name}.json`), ENCODING);
		return JSON.parse(contents) as T;
	}

	/**
	 * Check if a file exists in the current manager space
	 */
	protected async fileExists(name: string) {
		invariant(this.prepare, `Ensure to call "this.prepare" in ${this.name}.init() implementation`);
		return await fileExists(name);
	}

	/**
	 * Check if a folder exists in the current manager space
	 */
	protected async folderExists(name: string) {
		invariant(this.prepare, `Ensure to call "this.prepare" in ${this.name}.init() implementation`);
		return await folderExists(name);
	}
}
