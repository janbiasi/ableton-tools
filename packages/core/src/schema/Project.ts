import { IDatabaseEntry } from '../Database';

export interface IProjectSchema {
	name: string;
	path: string;
	sets: string[];
}

export interface IProjectBackupSchema {
	name: string;
	description: string;
	ref: string; // reference to the project ID
	version: number;
	path: string;
}

export interface IProjectMetaSchema {
	timestamp: IDatabaseEntry<IProjectSchema>['timestamp'];
	name: IProjectSchema['name'];
	ref: IDatabaseEntry<IProjectSchema>['id'];
	sets: IProjectSchema['sets'];
	template?: string;
}
