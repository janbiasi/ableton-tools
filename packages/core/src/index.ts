import { ENCODING, ABLETON_TOOLS_ROOT } from './const';
import { IProjectSchema, IProjectBackupSchema, IProjectMetaSchema } from './schema/Project';
import { ITemplateSchema } from './schema/Template';
import { ProjectManager, PROJECT_MANAGER_META_FILE } from './manager/ProjectManager';
import { TemplateManager } from './manager';
import { Runtime } from './Runtime';
import { Database } from './Database';
import {
	BackupRestorationError,
	InvariantViolationError,
	ProjectNotFoundError,
	ProjectMetaNotFoundError,
	DatabaseError,
} from './errors';
import {
	findLiveSetsInFolder,
	createArchive,
	unpackArchive,
	fileExists,
	folderExists,
	getLastPathFragment,
	invariant,
	safeDirName,
	formatBytes,
	uuid,
} from './utils';

export {
	// typedefs
	ITemplateSchema,
	IProjectSchema,
	IProjectBackupSchema,
	IProjectMetaSchema,
	// constants
	ENCODING,
	ABLETON_TOOLS_ROOT,
	PROJECT_MANAGER_META_FILE,
	// core
	Runtime,
	Database,
	ProjectManager,
	TemplateManager,
	// utilities
	findLiveSetsInFolder,
	createArchive,
	unpackArchive,
	fileExists,
	folderExists,
	getLastPathFragment,
	invariant,
	safeDirName,
	formatBytes,
	uuid,
	// errors
	BackupRestorationError,
	InvariantViolationError,
	ProjectNotFoundError,
	ProjectMetaNotFoundError,
	DatabaseError,
};
