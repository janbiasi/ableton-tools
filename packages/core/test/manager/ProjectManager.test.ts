import { resolve } from 'path';
import { mocked } from 'ts-jest/utils';
import { IProjectSchema } from '../../src/schema/Project';
import { ProjectManager } from '../../src/manager';
import { TEMP_TARGET_DIR, FIXTURES_DIR } from '../const';
import { toSnapshot } from './toSnapshot';

jest.mock('fs-extra');
jest.mock('globby');
import globby from 'globby';
import * as fs from 'fs-extra';
import { ITestProjectManager } from './TestManager';

const fsMock = mocked(fs);

(fsMock.lstat as any).mockReturnValue(
	Promise.resolve({
		isFile: () => false,
		isDirectory: () => false,
	})
);

describe('Core > Manager > ProjectManager', () => {
	beforeEach(() => {
		jest.clearAllMocks();
	});

	it('should be instanciable without options', () => {
		const pm = new ProjectManager();
		expect(toSnapshot(pm)).toMatchSnapshot();
	});

	it('should create base folders on prepare', async () => {
		const pm = new ProjectManager('/test');
		await pm.init();

		expect(fsMock.mkdirp).toHaveBeenCalledTimes(4);
		expect((fsMock.mkdirp as any).mock.calls).toEqual([
			['/test/projects'],
			['/test'],
			['/test'], // TODO: figure out why twice /test!
			['/test/projects/backups'],
		]);
	});

	describe('init', () => {
		it('should create databases if not available', async () => {
			const pm = new ProjectManager('/test');
			await pm.init();

			expect(fsMock.readFile).toHaveBeenCalledTimes(0);
			expect(fsMock.writeFile).toHaveBeenCalledTimes(2);
		});

		it('should write empty dbs on sync if none rehydrated', async () => {
			const pm = new ProjectManager('/test');
			await pm.init();

			expect(fsMock.readFile).toHaveBeenCalledTimes(0);
			expect(fsMock.writeFile).toHaveBeenCalledTimes(2);
			expect(fsMock.writeFile).toHaveBeenLastCalledWith(
				'/test/backups.json',
				JSON.stringify({ name: 'backups', entries: [] }, null, 2)
			);
		});

		it('should read from the existing database if available', async () => {
			(fsMock.lstat as any).mockReturnValueOnce(
				Promise.resolve({
					isFile: () => true,
					isDirectory: () => false,
				})
			);
			(fsMock.readFile as any).mockReturnValueOnce(
				Promise.resolve(
					JSON.stringify({
						entries: [
							{
								data: {
									name: 'test-template',
								},
							},
						],
					})
				)
			);
			const pm = new ProjectManager('/test');
			const registry = ((pm as unknown) as ITestProjectManager).registry;
			await pm.init();

			expect(fsMock.readFile).toHaveBeenCalledTimes(1);
			expect(fsMock.readFile).toHaveBeenLastCalledWith('/test/registry.json', 'utf8');
			expect(registry.contents.entries).toHaveLength(1);
			expect(registry.contents.entries[0].data.name).toEqual('test-template');
		});
	});

	describe('interaction', () => {
		(globby as any).mockReturnValue(Promise.resolve([]));

		it('should save templates correctly', async () => {
			const pm = new ProjectManager('/test');
			await pm.init();
			await pm.save({
				name: 'project',
				path: '/some/project/path',
			});

			const templates = ((pm as unknown) as ITestProjectManager).registry.contents.entries;
			expect(templates).toHaveLength(1);
			expect(templates[0]!.id).toBeTruthy();
			expect(templates[0]!.timestamp).toBeTruthy();
			expect(templates[0]!.data.name).toEqual('project');
			expect(templates[0]!.data.path).toEqual('/some/project/path');
			expect(templates[0]!.data.sets).toHaveLength(0);
		});
	});

	describe('livesets', () => {
		it('should find live sets correctly (mocked glob)', async () => {
			(globby as any).mockReturnValue(Promise.resolve(['/some/project/path/MySet.als']));
			const pm = new ProjectManager('/test');
			await pm.init();
			await pm.save({
				name: 'project',
				path: '/some/project/path',
			});

			const foundProject = pm.getByName('project');
			if (!foundProject) {
				fail('No project found');
			}

			expect(foundProject).toBeTruthy();
			expect(foundProject!.data.path).toEqual('/some/project/path');
			expect(foundProject!.data.sets).toHaveLength(1);
			expect(foundProject!.data!.sets![0]).toEqual('/some/project/path/MySet.als');
		});
	});
});
