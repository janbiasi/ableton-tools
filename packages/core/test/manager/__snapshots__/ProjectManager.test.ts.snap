// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Core > Manager > ProjectManager should be instanciable without options 1`] = `
Object {
  "backups": Database {
    "config": Object {
      "fileName": "backups",
      "name": "backups",
      "root": ".",
    },
    "contents": Object {
      "entries": Array [],
      "name": "unknown",
    },
  },
  "events": EventEmitter {
    "_events": Object {},
    "_eventsCount": 0,
    "_maxListeners": undefined,
  },
  "name": "projects",
  "prepared": false,
  "registry": Database {
    "config": Object {
      "fileName": "registry",
      "name": "projects",
      "root": ".",
    },
    "contents": Object {
      "entries": Array [],
      "name": "unknown",
    },
  },
  "root": "root://",
}
`;
